app/models/category.rb

```diff
@@ -227,6 +225,32 @@ class Category < ActiveRecord::Base
   # Allows us to skip creating the category definition topic in tests.
   attr_accessor :skip_category_definition
 
+  def self.preload_user_fields!(guardian, categories)
+    # Load notification levels
+    notification_levels = CategoryUser.notification_levels_for(guardian.user)
+    notification_levels.default = CategoryUser.default_notification_level
+
+    # Load permissions
+    allowed_topic_create_ids =
+      if !guardian.is_admin? && !guardian.is_anonymous?
+        Category.topic_create_allowed(guardian).where(id: categories.map(&:id)).pluck(:id).to_set
+      end
+
+    # Categories with children
+    with_children =
+      Category.where(parent_category_id: categories.map(&:id)).pluck(:parent_category_id).to_set
+
+    # Update category attributes
+    categories.each do |category|
+      category.notification_level = notification_levels[category[:id]]
+
+      category.permission = CategoryGroup.permission_types[:full] if guardian.is_admin? ||
+        allowed_topic_create_ids&.include?(category[:id])
+
+      category.has_children = with_children.include?(category[:id])
+    end
```

Comment:
It looks like this is duplicating logic in `Site#categories`. Perhaps we should try to avoid this duplication to prevent the two implementation from drifting?

Reply:
Yes and no. It is a bit different because here we use actual `Category` objects, while in `Site#categories` serialized objects are used instead.

Reply:
Hmm if that is the case let’s not tackle this for now. We can consider doing it in a future PR.